yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala

```diff
@@ -318,6 +314,13 @@ private[spark] class Client(
     env("SPARK_YARN_STAGING_DIR") = stagingDir
     env("SPARK_USER") = UserGroupInformation.getCurrentUser().getShortUserName()
 
+    // Propagate SPARK_HOME to the containers. This is needed for pyspark to
+    // work, since the executor's PYTHONPATH is built based on the location
+    // of SPARK_HOME.
+    sparkConf.getOption("spark.home").orElse(sys.env.get("SPARK_HOME")).foreach { path =>
+      env("SPARK_HOME") = path
+    }
```

Comment:
I think I figured this one out.

The assembly generated by maven includes the python files needed to run pyspark. The assembly generated by sbt doesn't. So if you deploy the sbt assembly, since `SPARK_HOME` is not propagated, it won't include the py4j and pyspark modules in `PYTHONPATH`, and pyspark jobs will fail.

In any case, unrelated to this issue and this is probably the wrong fix anyway, so I'll revert this part.